### PR TITLE
Components: Type BaseControl and VisuallyHidden

### PR DESCRIPTION
**Pull request by sirreal**
## Description

- Type components `utils`
- Type `BaseControl`
- Type `VisuallyHidden`

Builds on https://github.com/WordPress/gutenberg/pull/26066
Part of #18838 

## How has this been tested?

Type generation completes successfully.

## Types of changes

Internal / Code Quality: Add Type Checking to components.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/base-control/index.js [hunk 1]
```diff
@@ -8,6 +8,20 @@ import classnames from 'classnames';
  */
 import VisuallyHidden from '../visually-hidden';
 
+/**
+ * @typedef Props
```
Comment (youknowriad) [inline packages/components/src/base-control/index.js:12]: is this type "local"? I mean the name is very generic, is it fine?

Comment (sirreal) [reply packages/components/src/base-control/index.js:12]: It is local to the module, although it is an exported type. It's pretty common practice to name types like this in my experience.

File: packages/components/src/base-control/index.js [hunk 2]
```diff
@@ -8,6 +8,20 @@ import classnames from 'classnames';
...
+ * @property {string} id ID
+ * @property {import('react').ReactNode} help Help text
+ * @property {import('react').ReactNode} label Help text
```
Comment (sirreal) [inline packages/components/src/base-control/index.js:15]: Should be Label…

It's tricky to name `ReactNode` properties because they can be React elements, strings…

Maybe Label content?

**Comment by sirreal** [comment]
I've reverted to a draft. `BaseControls` relies on another module after #25842 which will need to be typed.

**Comment by sirreal** [comment]
I typed utils and this is ready for review again.

**Comment by sirreal** [comment]
I seem to have made a bunch of tests fail so I'll look again 🙂

File: packages/block-editor/src/components/responsive-block-control/test/__snapshots__/index.js.snap [hunk 3]
```diff
@@ -1,3 +1,3 @@
...
...y=\\"rbc-desc-0\\">All</span><span class=\\"components-visually-hidden\\" id=\\"rbc-desc-0\\">Controls the padding property for All viewports.</span></label></div><div class=\\"components-input-control__container css-nj8rmx-Container e1cr7zh11\\"><select class=\\"components-select-control__input css-owzv2l-Select e12x0a390\\" id=\\"inspector-select-control-0\\"><option value=\\"\\">Please select</option><option value=\\"small\\">Small</option><option value=\\"medium\\">Medium</option><option value=\\"large\\">Large</option></select><span class=\\"components-input-control__suffix css-1lym30p-Suffix e1cr7zh17\\"><div class=\\"css-s55r1c-DownArrowWrapper e12x0a391\\"><svg viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" role=\\"img\\" aria-hidden=\\"true\\" focusable=\\"false\\"><path d=\\"M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z\\"></path></svg></div></span><div aria-hidden=\\"true\\" class=\\"components-input-control__backdrop css-1o1vxqx-BackdropUI e1cr7zh15\\"></div></div></div></div></div><p id=\\"all\\">All is used here for testing purposes to ensure we have access to details about the device.</p></div></div></fieldset>"`;
```
Comment (sirreal) [inline packages/block-editor/src/components/responsive-block-control/test/__snapshots__/index.js.snap:3]: The diff here is like:

- `css-hwcz8s-Select` -> `css-owzv2l-Select`
- `css-1s1n1to-BackdropUI` -> `css-1o1vxqx-BackdropUI`

Presumably this is because some of the files involved in calculating the styles have changed and the hash is updated. @ItsJonQ @saramarcondes does this seem alright? Is there a better way to snapshot styles like this so they're stable?

File: packages/components/src/base-control/index.js [hunk 4]
```diff
@@ -14,6 +14,26 @@ import {
 	StyledHelp,
 } from './styles/base-control-styles';
 
+/**
+ * @typedef Props
+ * @property {string} id The id of the element to which labels and help text are being generated.
+ *                       That element should be passed as a child.
+ * @property {import('react').ReactNode} help If this property is added, a help text will be
```
Comment (saramarcondes) [inline packages/components/src/base-control/index.js:21]: Out of curiosity (rather than thinking it _should_ be done) why don't we alias this type like we do the rest of them in the `@wordpress/element` package?

File: packages/components/src/visually-hidden/index.js [hunk 5]
```diff
@@ -8,13 +8,25 @@ import classnames from 'classnames';
...
+ * @template {keyof JSX.IntrinsicElements | import('react').JSXElementConstructor<any>} T
+ * @typedef OwnProps
+ * @property {T} [as='div'] Component to render, e.g. `"div"` or `MyComponent`.
+ */
+
+/**
+ * @template {keyof JSX.IntrinsicElements | import('react').JSXElementConstructor<any>} T
+ * @typedef {OwnProps<T> & import('react').ComponentProps<T>} Props
+ */
+
```
Comment (saramarcondes) [inline packages/components/src/visually-hidden/index.js:21]: Nice! I was taking a look at this last night and having a hard time wrapping my head around how to type this. Good work!

File: packages/block-editor/src/components/responsive-block-control/test/__snapshots__/index.js.snap (hunk 3 above)
Comment (saramarcondes) [reply packages/block-editor/src/components/responsive-block-control/test/__snapshots__/index.js.snap:3]: Honestly a better question is whether this snapshot test is needed at all:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/responsive-block-control/test/index.js#L69-L121

We're already testing for a bunch of assertions inside the container's markup, it's not clear to me, as someone who's only just looking at these tests, why the snapshot is valuable (especially considering we're using `render` which makes the snapshot nearly impossible to sanely diff).

On the other hand, you didn't change anything about the styles, I have no idea why the class names that emotion generates wouldn't be idempotent.

File: packages/components/src/base-control/index.js (hunk 4 above)
Comment (sirreal) [reply packages/components/src/base-control/index.js:21]: You're referring to how we alias e.g. `ReactElement` as `WPElement`? I don't see any value in that and would prefer to use the types we're referring to more directly.

Additionally, a lot of the aliases lose the type parameters, which is unfortunate.

That's the only _why_ for this and is very much tied to the question of whether or not we _should_ 🙃

Comment (saramarcondes) [reply packages/components/src/base-control/index.js:21]: > Additionally, a lot of the aliases lose the type parameters, which is unfortunate.

I noticed too that when I tried to alias `ReactNode` it lost some of the type itself for some reason (it was no longer able to be `null` for example). Do you know why that is?

Comment (sirreal) [reply packages/components/src/base-control/index.js:21]: I don't understand why that one would be affected. The problem with type parameters is that we don't provide them in the aliases at the moment. We'd need to rewrite them as

```diff
diff --git a/packages/element/src/react.js b/packages/element/src/react.js
index 520401dcc5..2bb42c77b4 100644
--- a/packages/element/src/react.js
+++ b/packages/element/src/react.js
@@ -37,7 +37,8 @@ import { isString } from 'lodash';
 /**
  * Object containing a React component.
  *
- * @typedef {import('react').ComponentType} WPComponent
+ * @template P
+ * @typedef {import('react').ComponentType<P>} WPComponent
  */
 
 /**
```

This _also_ isn't great because now we have a required type parameter. I don't think there's a good way to write type parameter (@template) defaults in JSDoc at this time.

File: packages/components/src/utils/math.js [hunk 6]
```diff
@@ -19,26 +19,34 @@ export function getNumber( value ) {
...
-	return args.reduce( ( sum, arg ) => sum + getNumber( arg ), 0 );
+	return args.reduce(
+		/** @type {(sum:number, arg: number|string) => number} */
+		( sum, arg ) => sum + getNumber( arg ),
+		0
+	);
 }
 
 /**
  * Safely subtracts 2 values.
```
Comment (sirreal) [inline packages/components/src/utils/math.js:35]: @ItsJonQ Is this function helpful? Typically you don't see subtract methods on an arbitrary number of elements, probably because it's not an associative operation.

The implementation could be vastly simplified if this were actually designed to work on 2 arguments like the description suggests.

Looking at https://github.com/WordPress/gutenberg/pull/23006, it doesn't seem like there was a need for this to operate on an arbitrary number of elements.

Comment (ItsJonQ) [reply packages/components/src/utils/math.js:35]: @sirreal I think this function is really helpful 🙏 .

Lately, I've been relying on these functions for `Input` related adjustments in G2.
The annoying thing about input values is they can be strings numeric values, or nulls, empty strings, etc...

Having something that can reliably add (assumed) numeric values together is really helpful.
The previous implementation only supported 2 values, so that description is out of date 🙈 .
Thank you for catching that.

P.S. If you know of a better way to reliably handle MATH operations for assumed number types... I'm all ears <3!

Comment (sirreal) [reply packages/components/src/utils/math.js:35]: > I think this function is really helpful 🙏 .

I'm on board with functions providing the utility we need, I wasn't questioning the utility of a subtraction function that can operate on numbers or number-like string values, I'll clarify below 😉 

> If you know of a better way to reliably handle MATH operations for assumed number types...

I'm in favor of keeping things small and dependencies minimal, but this module already depends on lodash and it could pull in a few more things. In the context of WordPress, that will actually reduce the bundle size because lodash will be available in its entirety.

Some of these are also very tricky tricky to get right and lodash is battle tested:

- `getNumber` -> [`_.toNumber`](https://lodash.com/docs/4.17.15#toNumber)
- `add` -> [`( values ) => _.sumBy( values, _.toNumber )`](https://lodash.com/docs/4.17.15#sumBy)

> Having something that can reliably add [any number of] (assumed) numeric values together is really helpful.

Yep, this is perfectly fine. I have no concerns about `add`. It's `subtract` that concerns me. This may seem pedantic, but bear with me 😅  I wonder specifically about subtraction of an arbitrary number of elements which seems strange. A variadic subtraction function (subtract an arbitrary number of values) is a bit odd. It's important that we subtract _from_ a number or [things get weird](https://www.typescriptlang.org/play?#code/PQKhCgAIUgFBDATgZwKbMvAdgE0o1AF0QEtUA3dTSLAVwFsAjVRSc+AG1tQDooJokAAIAHJPHqQA3rSwBrLAHsA7lgC+bTt0gAVABapIJLAGNF9YwHNNXXv37CChWoizS6TFhv2GxKVHgezKzstnzQwOAAZrImhCSKbpZEAHIMwQAUodwAXJDIxFaQAD406SwAlNLgAJBmWAVlnqwAvJBpzVlaqBUA3OC1Ti5uJMgp8CkZQZWQAPyQAAyQedOI-WoDoAIwAMrwUagcAJ75tIzE8HEYAEw23Mjh28J+EtIAgoiI8EcAPKvFBVIWEsAD4NEhLBgAGrdDCERSnc5fOKQeHJQgGRCPBxCIaudzlRDeAyQHAkKIHAimQyKKKoknZdCPSIxUzxRKIi5xDI8XkQ5B5D5fX6AoqlVYgqpSQZEYaYRCQngEHC0EyoDK1GoZMkUlaEgA08sseVFwJKTWChuMOFQAA8KnrmpAWiDqjV3fVGoznZB0R1MhC+pqaniRrg7c6Wm0lvNvXkdXSALR3VD9d1qfWaha1IMbaKxdlueA4HA8vkKgWQIXfH6m6ziwmSt2ho0PZWq9Wa7XkqKOy1Gk2FM0N5pW8P2vssZ2u6XuuqJL3dH1+wkZQNp90thOQADUKY3GazOfWm2AukxqAA5BhkOYiHoisgRKgTCROJhGIpaIROcjCzxIAAdUMLBUACVEEQKJAf3gGDIAAfQsOhUFweDIEUVgMSsDAcESS8f3oeA5EMNAGjsT1FA4XgOEUSwMmQM4uUIDIKgqcBgDPJZAF4NwAhPcgAB5ABpcAKKongaLohikUuZiAEYFlY9iz3kyAuMAVD3IAATVQeA9EAMgIRIXSjqNo+jGL-MzpO5CpDXkxSOMgRMVK4wAaPcgZRUBIRAcAM0STMk8yZMspjgos+TDSkkLWJsyA7LYhzuLcjyvJ809IAASTceoLgKQ1izJQsjAwAArWhGkuEwGFoDhYKKeCGPoQjcGQeDDIaYzxNM-KWPizjIEAUHI2tvMSJLXEsMjipTYqWQa-M6ujuu6mLursvLxpWhToqm64ZvAIA):

```ts
// There's something special about subtraction. We need to start at a _minuend_ or things don't make sense.
console.log(subtract())
// 0 🙂 OK
console.log(subtract(10))
// 10 😕 Yeah…
console.log(subtract(subtract(), 10))
// -10 😬 weird…
console.log(subtract(subtract(subtract(10, subtract())), 10))
// 0 😬 weird…

// In contrast, addition is just accumulating _summands_
console.log(add())
// 0 ✅
console.log(add(10))
// 10 ✅
console.log(add(add(), add(10), add(add(10))))
// 20 ✅
```

You might also consider that the result of subtraction the difference. What is the difference of 3 things?

If you look, you can see this detail buried in the current implementation. Look at line 40 here:

https://github.com/WordPress/gutenberg/blob/2cf608a2095f35f495f189cb2cfcd4f62631c247/packages/components/src/utils/math.js#L37-L42

There's an exception: don't subtract the first value, just pass it through. If we rethink this as a function that takes a minuend (our strating point) and we subtract a variadic number of subtrahends from it (0 or more) it makes more sense. Our function returns the difference of the minuend and the sum of the subtrahends:

```ts
function subtract( minuend: string|number, ...subtrahends: Array<string | number>): number {
	return getNumber( minuend ) - add( ...subtrahends );
}
```

So, this is a little bit pedantic, the result is the same, but it makes more sense to me. There's a difference (no pun intended) between the first value and the rest of the values that's significant.

Comment (sirreal) [reply packages/components/src/utils/math.js:35]: …this is also a bit off topic here. I'll open a PR with some of the suggested changes and the conversation can continue there.

**Comment by sirreal** [comment]
I've rebased and fixed the conflicts, this can be reviewed.

File: packages/components/src/utils/hooks/use-update-effect.js [hunk 7]
```diff
@@ -3,10 +3,13 @@
  */
 import { useRef, useEffect } from '@wordpress/element';
 
-/*
+/**
  * A `React.useEffect` that will not run on the first render.
  * Source:
  * https://github.com/reakit/reakit/blob/master/packages/reakit-utils/src/useUpdateEffect.ts
+ *
+ * @param {()=>void}                       effect
```
Comment (saramarcondes) [inline packages/components/src/utils/hooks/use-update-effect.js:11]: Nit: We could use the `EffectCallback` type from `React`

File: packages/components/src/utils/math.js [hunk 8]
```diff
@@ -6,7 +6,7 @@ import { clamp } from 'lodash';
 /**
  * Parses and retrieves a number value.
  *
- * @param {any} value The incoming value.
+ * @param {unknown} value The incoming value.
```
Comment (saramarcondes) [inline packages/components/src/utils/math.js:9]: Out of curiosity, why this change? I'm not fully up on when to use `unknown`. I guess it makes semantic sense here but what's the exact reason?

**Review by sarayourfriend — APPROVED** [review]
One nit pick but otherwise LGTM!

File: packages/components/src/utils/math.js (hunk 8 above)
Comment (sirreal) [reply packages/components/src/utils/math.js:9]: Glad you asked 🙂 

`any` and `unknown` are both boxes that we can put anything into. The difference is that `any` satisfies every other type, whereas `unknown` is the opposite. It satisfies no other type and requires us to be careful to use it. In code that's like:

```ts
// Anything _satisfies_ both types `any` and `unknown`
var __a: any = null
var __a: any = undefined
var __a: any = ['foo']
var __a: any = 0
const a: any = new Set(['apples', 'bananas'])

var __u: unknown = null
var __u: unknown = undefined
var __u: unknown = ['foo']
var __u: unknown = 0
const u: unknown = new Set(['apples', 'bananas'])


// Type system catches this. `unknown` doesn't satisfy anything! Phew…
u.toLocaleLowerCase(); // TYPE ERROR!
// We can't call set methods either, we said we didn't know anything >:/
u.has('foo'); // TYPE ERROR!
// Typically we'll use type guards to be safe with our `unknown` types :D
u instanceof Set && u.has('apples'); // No type error 👍

// `any` type satisfies everything. It's dangerous, as if we turn off the type system completely!!
a.toLocaleLowerCase() // Oops… Set doesn't implement that! We hit s runtime error "TypeError: u.toLocaleLowerCase is not a function" 💥
```

**Review by sarayourfriend — APPROVED** [review]

**Comment by ItsJonQ** [comment]
Haiii! I tested out a couple of the utils. A lot of this TypeScript is a bit advanced for me 😅 .
However, I'm able to get the expected Intellisense experience within VSCode as far as comments + type errors go :)

**Comment by ItsJonQ** [comment]
Following up with this one! It looks like the `package.json` needs updating. Hopefully a `rebase`/`merge` from `master` branch will make the tests happier 🙏

**Comment by sirreal** [comment]
Thanks for the feedback, everyone. Rebased to fix conflicts.

**Review by sarayourfriend — APPROVED** [review]

**Comment by sirreal** [comment]
Rebased, I think tests are back on track today and we can get this landed 🤞
